File: python/vdm_service/runtime.py
"""Application orchestration. MQTT callbacks persist only; workers perform IO/RPC."""
import hashlib
import json
import logging
import threading
import time
import httpx
from vdm_mqtt_sdk import VdmMqttClient, VdmMqttClientConfig, VdmCodec, VdmTopics, RpcError
from .store import Store
from .modules import alarms, device, measurement, targets
from .modules.evidence import EvidenceService, process_event

LOG = logging.getLogger(__name__)

class ServiceError(Exception):
    def __init__(self, status, code, message, **details):
        self.status = status
        self.body = {"error":{"code":code,"message":message,**details}}
        super().__init__(message)


class Service:
    def __init__(self, settings, *, client_factory=VdmMqttClient, webhook_sender=None):
        self.settings = settings
        self.store = Store(settings)
        self.evidence = EvidenceService(self.store,settings)
        self.devices = {(d['connectionId'],d['deviceId']):d for d in settings.devices}
        self.clients = {}
        self.stop_event = threading.Event()
        self.reconnect = set()
        self.reconnect_lock = threading.Lock()
        self.threads = []
        self.started = False
        self.closed = False
        self.webhook_sender = webhook_sender or self._send_webhook
        self.rpc_slots = threading.BoundedSemaphore(32)
        for (cid,did), spec in self.devices.items():
            digest = hashlib.sha256(json.dumps([cid,did],separators=(',',':')).encode()).hexdigest()[:32]
            config = VdmMqttClientConfig(host=spec['host'],port=spec['port'],topics=VdmTopics.for_device(did),
                payload_format=spec['format'],username=spec['username'],password=spec['password'],
                client_id='vdm-service-python-'+digest,persistent_session=True,manual_ack=True,
                subscription_suffixes=('telemetry','attributes','3A','event','image','rpc/resp'))
            self.clients[(cid,did)] = client_factory(config,
                on_raw_message=lambda topic,raw,c=cid,d=did:self.ingest(c,d,topic,raw),
                on_error=lambda exc,c=cid,d=did:self.transport_error(c,d,exc))

    def get_device(self,cid,did):
        if (cid,did) not in self.devices:
            raise ServiceError(404,'NOT_FOUND','unknown connection/device')
        return self.devices[(cid,did)]

    def ingest(self,cid,did,topic,payload):
        self.get_device(cid,did)
        try:
            self.store.enqueue(cid,did,topic,payload)
        except Exception:
            # Schedule reconnect even if writing a counter also fails (disk full).
            with self.reconnect_lock:
                self.reconnect.add((cid,did))
            try:
                self.store.increment('ingressFailures')
            except Exception:
                pass
            LOG.warning('durable ingress rejected for %s/%s',cid,did)
            raise

    def transport_error(self,cid,did,exc):
        LOG.warning('MQTT %s/%s: %s',cid,did,type(exc).__name__)
        if not self.stop_event.is_set():
            with self.reconnect_lock:
                self.reconnect.add((cid,did))
        try:
            self.store.increment('mqttErrors')
        except Exception:
            pass

    def process_one(self):
        row = self.store.next_pending('inbox')
        if row is None:
            return False
        cid,did = row['connection_id'],row['device_id']
        try:
            spec = self.get_device(cid,did)
            decoded = VdmCodec(spec['format']).decode(row['topic'],VdmTopics.for_device(did),bytes(row['payload']))
            if decoded.suffix == 'image':
                self.evidence.process(cid,did,decoded)
            with self.store.transaction() as db:
                if decoded.suffix == 'telemetry':
                    measurement.process(db,cid,did,decoded.as_dict())
                elif decoded.suffix == 'attributes':
                    device.process(db,cid,did,decoded.as_dict())
                elif decoded.suffix == '3A':
                    alarms.process(db,cid,did,decoded.as_dict(),self.settings.notification_max_age_seconds,self.settings.max_inbox_rows)
                elif decoded.suffix == 'event':
                    process_event(db,cid,did,decoded.as_dict())
                db.execute("UPDATE inbox SET status='done',error=NULL WHERE id=?",(row['id'],))
        except Exception as exc:
            self.store.fail('inbox',row,exc)
            LOG.warning('inbox processing failed id=%s type=%s',row['id'],type(exc).__name__)
        return True

    def invoke(self,cid,did,operation,params):
        spec = self.get_device(cid,did)
        capability = operation['capability']
        if capability != '-' and capability not in spec['capabilities']:
            raise ServiceError(409,'UNSUPPORTED_CAPABILITY',f'operation requires {capability}')
        method = operation['method']
        client = self.clients[(cid,did)]
        if method == 'ackEvidencePackage' and not self.evidence.verified(cid,did,params.get('eventId'),params.get('packageSha256')):
            raise ServiceError(400,'INVALID_ARGUMENT','evidence must be received and verified before ACK')
        try:
            VdmCodec(spec['format']).encode_rpc_request(method,params,1)
        except (TypeError,ValueError) as exc:
            raise ServiceError(400,'INVALID_ARGUMENT',str(exc)) from None
        if not client.is_connected:
            raise ServiceError(503,'UNAVAILABLE','MQTT device connection is unavailable')
        if not self.rpc_slots.acquire(blocking=False):
            raise ServiceError(503,'OVERLOADED','RPC concurrency limit reached')
        try:
            result = (targets.invoke(client,method,params,self.settings.rpc_timeout_ms/1000)
                      if operation['module']=='targets' else
                      client.call(method,params,timeout=self.settings.rpc_timeout_ms/1000))
            response = result.as_dict()
            # Check again for adapters and SDKs configured to return device errors.
            _,code,_,_ = VdmCodec(spec['format']).response_info(getattr(result, 'value', response))
            if code:
                raise ServiceError(502,'DEVICE_ERROR','device rejected request',deviceCode=code)
            return {"connectionId":cid,"deviceId":did,"method":method,
                    "status":"accepted" if operation['mode']=='async' else 'completed',"response":response}
        except ServiceError:
            raise
        except RpcError as exc:
            raise ServiceError(502,'DEVICE_ERROR','device rejected request',deviceCode=exc.code) from None
        except TimeoutError:
            raise ServiceError(504,'RPC_TIMEOUT','device response timed out',outcome='unknown') from None
        except (ValueError,TypeError):
            raise ServiceError(400,'INVALID_ARGUMENT','invalid device request parameters') from None
        except (ConnectionError,OSError,RuntimeError):
            raise ServiceError(503,'UNAVAILABLE','MQTT request failed') from None
        finally:
            self.rpc_slots.release()

    def _call_worker(self,cid,did,method,params):
        client = self.clients[(cid,did)]
        if not client.is_connected:
            raise ConnectionError('device connection unavailable')
        result = client.call(method,params,timeout=self.settings.rpc_timeout_ms/1000)
        response = result.as_dict()
        _,code,_,_ = VdmCodec(self.devices[(cid,did)]['format']).response_info(getattr(result, 'value', response))
        if code:
            raise RpcError(0,code,'worker RPC rejected')
        return response

    def process_job(self):
        row = self.store.next_pending('jobs')
        if row is None:
            return False
        cid,did = row['connection_id'],row['device_id']
        try:
            with self.store.lock:
                watermark = self.store.db.execute('SELECT coalesce(max(rowid),0) FROM alarm_events WHERE connection_id=? AND device_id=?',(cid,did)).fetchone()[0]
            response = self._call_worker(cid,did,'getAlarmState',{})
            with self.store.transaction() as db:
                current = db.execute('SELECT coalesce(max(rowid),0) FROM alarm_events WHERE connection_id=? AND device_id=?',(cid,did)).fetchone()[0]
                if current != watermark:
                    # A live event raced the snapshot request; don't overwrite it.
                    db.execute('UPDATE jobs SET due=? WHERE key=?',(time.time()+1,row['key']))
                    return True
                alarms.reconcile(db,cid,did,response)
                db.execute("UPDATE jobs SET status='done' WHERE key=?",(row['key'],))
        except Exception as exc:
            self.store.fail('jobs',row,exc)
        return True

    def process_ack(self):
        row = self.store.next_pending('evidence')
        if row is None:
            return False
        cid,did = row['connection_id'],row['device_id']
        try:
            if not self.evidence.verified(cid,did,row['event_id'],row['sha']):
                raise ValueError('verified evidence receipt no longer valid')
            kind = 'EVIDENCE_KIND_SNAPSHOT' if self.devices[(cid,did)]['format']=='protobuf' else 'SNAPSHOT'
            self._call_worker(cid,did,'ackEvidencePackage',dict(eventId=row['event_id'],kind=kind,packageSha256=row['sha']))
            with self.store.transaction() as db:
                db.execute("UPDATE evidence SET status='done',error=NULL WHERE key=?",(row['key'],))
        except Exception as exc:
            self.store.fail('evidence',row,exc)
        return True

    def _send_webhook(self,url,payload,headers):
        # No redirects: do not forward bearer secrets to another host.
        with httpx.Client(timeout=5.0,follow_redirects=False,trust_env=False) as client:
            response = client.post(url,json=payload,headers=headers)
            if not 200 <= response.status_code < 300:
                raise RuntimeError('webhook rejected delivery')

    def process_outbox(self):
        if not self.settings.webhook_url:
            return False
        row = self.store.next_pending('outbox')
        if row is None:
            return False
        headers = {'Idempotency-Key':row['key']}
        if self.settings.webhook_token:
            headers['Authorization']='Bearer '+self.settings.webhook_token
        try:
            self.webhook_sender(self.settings.webhook_url,json.loads(row['payload']),headers)
            with self.store.transaction() as db:
                db.execute("UPDATE outbox SET status='done',error=NULL WHERE key=?",(row['key'],))
        except Exception as exc:
            self.store.fail('outbox',row,exc)
        return True

    def health(self):
        connections = [dict(connectionId=c,deviceId=d,connected=bool(client.is_connected)) for (c,d),client in self.clients.items()]
        counters = self.store.counters()
        degraded = any(not item['connected'] for item in connections) or any(v for k,v in counters.items() if k.endswith('_failed'))
        return dict(status='degraded' if degraded else 'ok',connections=connections,counters=counters)

    def _loop(self,work):
        while not self.stop_event.is_set():
            try:
                if work():
                    continue
            except Exception as exc:
                LOG.error('worker failure: %s',type(exc).__name__)
            self.stop_event.wait(.1)

    def prune(self):
        self.evidence.prune()
        self.store.prune()

    def _maintenance(self):
        while not self.stop_event.wait(1):
            with self.reconnect_lock:
                reconnect,self.reconnect = self.reconnect,set()
            for key in reconnect:
                try:
                    self.clients[key].reconnect_background()
                except Exception as exc:
                    self.transport_error(*key,exc)
                    with self.reconnect_lock:
                        self.reconnect.add(key)
            try:
                self.prune()
            except Exception as exc:
                LOG.error('retention failed: %s',type(exc).__name__)

    def start(self):
        if self.started:
            return
        self.started = True
        for key,client in self.clients.items():
            try:
                client.start_background()
            except Exception as exc:
                self.transport_error(*key,exc)
                with self.reconnect_lock:
                    self.reconnect.add(key)
        for work in (self.process_one,self.process_job,self.process_ack,self.process_outbox):
            thread = threading.Thread(target=self._loop,args=(work,),daemon=True)
            thread.start(); self.threads.append(thread)
        thread = threading.Thread(target=self._maintenance,daemon=True)
        thread.start(); self.threads.append(thread)

    def close(self):
        if self.closed:
            return
        self.closed = True
        self.stop_event.set()
        for client in self.clients.values():
            client.stop()
        for thread in self.threads:
            thread.join(timeout=self.settings.rpc_timeout_ms/1000+6)
        self.store.close()

File: javascript/service/runtime.js
"use strict";
const path = require("node:path");
const fs = require("node:fs");
const crypto = require("node:crypto");
const { Worker } = require("node:worker_threads");
const { VdmMqttClient, VdmTopics, RpcError, VdmCodec } = require("../sdk");
const { Store, digest } = require("./store");
const { fail } = require("./app");
const { consumeImage } = require("./modules/evidence");
class Runtime {
  constructor(config, directory) {
    this.config = config;
    this.directory = directory;
    this.store = new Store(directory, config);
    this.clients = new Map();
    this.stopping = false;
    this.busy = false;
    this.activeCalls = 0;
    this.imageWorker = new Worker(path.join(__dirname, "evidence-worker.js"), {
      workerData: {
        directory: path.join(directory, "evidence"),
        maxBytes: config.maxEvidenceBytes,
      },
    });
    this.imageRequests = new Map();
    this.imageSequence = 0;
    this.imageWorker.on("message", (m) => {
      const p = this.imageRequests.get(m.id);
      this.imageRequests.delete(m.id);
      if (p) m.error ? p.reject(new Error(m.error)) : p.resolve(m.completed);
    });
    this.imageWorker.on("error", (e) => {
      for (const p of this.imageRequests.values()) p.reject(e);
      this.imageRequests.clear();
      this.log(e);
    });
    for (const c of config.connections)
      for (const d of c.devices) {
        const key = JSON.stringify([c.id, d.id]);
        const sdk = new VdmMqttClient(
          {
            host: c.host,
            port: c.port,
            topics: VdmTopics.forDevice(d.id),
            payloadFormat: d.format,
            username: process.env[c.usernameEnv || "MQTT_USERNAME"],
            password: process.env[c.passwordEnv || "MQTT_PASSWORD"],
            clientId: "vdm-js-" + digest([c.id, d.id]).slice(0, 32),
            subscriptionSuffixes: [
              "rpc/resp",
              "telemetry",
              "attributes",
              "3A",
              "event",
              "image",
            ],
            connectTimeoutMs: 3000,
            durableReceive: (topic, raw, qos) => {
              this.store.receive(
                c.id,
                d.id,
                VdmTopics.forDevice(d.id).suffix(topic),
                raw,
                d.format,
              );
              if (qos === 0) this.store.count("qos0Received");
            },
          },
          null,
          (e) => this.log(e),
        );
        this.clients.set(key, {
          sdk,
          connectionId: c.id,
          deviceId: d.id,
          format: d.format,
          capabilities: d.capabilities,
          connecting: false,
        });
      }
  }
  log(e) {
    console.error("SERVICE_ERROR", e.message);
    try {
      this.store.count("errors");
    } catch {
      /* database failure already logged */
    }
  }
  devices() {
    return [...this.clients.values()].map(
      ({ connectionId, deviceId, format, capabilities }) => ({
        connectionId,
        deviceId,
        format,
        capabilities,
      }),
    );
  }
  health() {
    const connections = [...this.clients.values()].map((x) => ({
      connectionId: x.connectionId,
      deviceId: x.deviceId,
      connected: Boolean(x.sdk.client?.connected),
    }));
    const counters = this.store.healthCounts();
    return {
      status:
        connections.every((x) => x.connected) &&
        !counters.inboxfailed &&
        !counters.jobsfailed &&
        !counters.outboxfailed
          ? "ok"
          : "degraded",
      connections,
      counters,
    };
  }
  async call(c, d, operation, params = {}) {
    const entry = this.clients.get(JSON.stringify([c, d]));
    if (!entry) throw fail(404, "NOT_FOUND", "Unknown connection/device");
    if (
      operation.method === "ackEvidencePackage" &&
      !this.store.verified(c, d, params)
    )
      throw fail(
        400,
        "INVALID_ARGUMENT",
        "Matching verified local evidence receipt required",
      );
    if (!entry.sdk.client?.connected)
      throw fail(503, "UNAVAILABLE", "MQTT connection unavailable");
    if (this.activeCalls >= 64)
      throw fail(503, "OVERLOADED", "RPC concurrency limit reached");
    try {
      entry.sdk.codec.encodeRpcRequest(operation.method, params, 1);
    } catch (e) {
      throw fail(400, "INVALID_ARGUMENT", e.message);
    }
    this.activeCalls++;
    try {
      return (
        await entry.sdk.call(operation.method, params, {
          timeoutMs: this.config.rpcTimeoutMs,
        })
      ).data;
    } catch (e) {
      if (e instanceof RpcError)
        throw fail(502, "DEVICE_ERROR", e.message, { deviceCode: e.code });
      if (/超时|timeout/i.test(e.message))
        throw fail(
          504,
          "RPC_TIMEOUT",
          "RPC result unknown; read back device state before retrying",
          { outcome: "unknown" },
        );
      throw fail(503, "UNAVAILABLE", "RPC connection lost; result unknown", {
        outcome: "unknown",
      });
    } finally {
      this.activeCalls--;
    }
  }
  image(c, d, raw) {
    const id = ++this.imageSequence;
    return new Promise((resolve, reject) => {
      this.imageRequests.set(id, { resolve, reject });
      this.imageWorker.postMessage({
        id,
        c,
        d,
        raw,
        persistentBytes: this.store.db
          .prepare("SELECT coalesce(sum(length(raw)),0) n FROM chunks")
          .get().n,
      });
    });
  }
  async start() {
    // Rebuild the assembler bitmap from committed raw chunks before receiving new
    // work. QoS1 chunks remain durable across restart, independent of MQTT session.
    for (const row of this.store.db
      .prepare("SELECT * FROM chunks ORDER BY c,d,event,idx")
      .iterate()) {
      try {
        const result = await this.image(row.c, row.d, row.raw);
        if (result) this.store.job(row.c, row.d, "ack", result);
      } catch (e) {
        this.log(e);
      }
    }
    this.connectTimer = setInterval(() => this.connect(), 1000);
    this.connect();
    this.workerTimer = setInterval(() => this.tick(), 100);
    this.pruneTimer = setInterval(() => {
      if (this.busy) return;
      try {
        this.store.prune();
        this.pruneEvidence();
      } catch (e) {
        this.log(e);
      }
    }, 60000);
  }
  connect() {
    for (const entry of this.clients.values())
      if (!entry.sdk.client && !entry.connecting && !this.stopping) {
        entry.connecting = true;
        entry.sdk
          .start()
          .catch((e) => this.log(e))
          .finally(() => {
            entry.connecting = false;
          });
      }
  }
  async tick() {
    if (this.busy || this.stopping) return;
    this.busy = true;
    try {
      for (const table of ["inbox", "jobs", "outbox"]) {
        const row = this.store.due(table);
        if (row) {
          try {
            if (table === "inbox") await this.consume(row);
            else if (table === "jobs") await this.job(row);
            else {
              if (!process.env.VDM_WEBHOOK_URL) continue;
              await this.notify(row);
            }
            this.store.done(table, row.id);
          } catch (e) {
            this.store.fail(table, row, e);
            this.log(e);
          }
        }
      }
    } catch (e) {
      this.log(e);
    } finally {
      this.busy = false;
    }
  }
  async consume(row) {
    const topics = VdmTopics.forDevice(row.d);
    const message = new VdmCodec(row.format).decode(
      topics.topic(row.suffix),
      topics,
      row.raw,
    );
    if (["telemetry", "attributes"].includes(row.suffix))
      this.store.db.transaction(() => {
        this.store.saveLatest(row.c, row.d, row.suffix, message.data);
        this.store.done("inbox", row.id);
      })();
    else if (row.suffix === "3A")
      this.store.alarm(
        row.c,
        row.d,
        message.data,
        Date.now() / 1000,
        this.config.notificationMaxAgeSeconds,
      );
    else if (row.suffix === "event")
      this.store.evidenceEvent(row.c, row.d, message.data);
    else if (row.suffix === "image") {
      await consumeImage(this, row, message);
    }
  }
  async job(row) {
    const p = JSON.parse(row.payload);
    if (row.kind === "ack") {
      const entry = this.clients.get(JSON.stringify([row.c, row.d]));
      const kind =
        entry.format === "json" ? "SNAPSHOT" : "EVIDENCE_KIND_SNAPSHOT";
      await this.call(
        row.c,
        row.d,
        { method: "ackEvidencePackage" },
        { eventId: p.eventId, kind, packageSha256: p.packageSha256 },
      );
      this.store.db
        .prepare("DELETE FROM chunks WHERE c=? AND d=? AND event=? AND hash=?")
        .run(row.c, row.d, String(p.eventId), p.packageSha256);
    } else if (row.kind === "reconcile") {
      const response = await this.call(
        row.c,
        row.d,
        { method: "getAlarmState" },
        {},
      );
      // Store authoritative snapshot separately. Keep ambiguous event state marked:
      // a state read cannot invent an ordering for equal timestamps.
      this.store.reconcile(row.c, row.d, response);
    }
  }
  async notify(row) {
    const response = await fetch(process.env.VDM_WEBHOOK_URL, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "Idempotency-Key": row.id,
        ...(process.env.VDM_WEBHOOK_TOKEN
          ? { Authorization: `Bearer ${process.env.VDM_WEBHOOK_TOKEN}` }
          : {}),
      },
      body: row.payload,
      signal: AbortSignal.timeout(5000),
      redirect: "error",
    });
    await response.body?.cancel();
    if (!response.ok) throw new Error(`webhook HTTP ${response.status}`);
  }
  pruneEvidence() {
    const base = path.join(this.directory, "evidence");
    const cutoff = Date.now() - this.config.retentionDays * 86400000;
    const walk = (dir) => {
      if (!fs.existsSync(dir)) return;
      for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
        const file = path.join(dir, ent.name);
        if (ent.isDirectory()) walk(file);
        else if (
          !ent.name.endsWith(".part") &&
          fs.statSync(file).mtimeMs < cutoff
        ) {
          const pending = this.store.db
            .prepare(
              "SELECT 1 FROM jobs WHERE kind='ack' AND status!='done' AND json_extract(payload,'$.packagePath')=?",
            )
            .get(file);
          if (!pending) fs.rmSync(file);
        }
      }
    };
    walk(base);
  }
  async stop() {
    this.stopping = true;
    clearInterval(this.connectTimer);
    clearInterval(this.workerTimer);
    clearInterval(this.pruneTimer);
    await Promise.allSettled(
      [...this.clients.values()].map((x) => x.sdk.stop()),
    );
    while (this.busy) await new Promise((r) => setTimeout(r, 20));
    await this.imageWorker.terminate();
    this.store.close();
  }
}
module.exports = { Runtime };
